File: test/py-stub-result-overlay/BaboonTests/custom_containers.py
from BaboonDefinitions.Generated.testpkg.pkg0.BaboonServiceRt import IBaboonServiceRt


class ResultOk:
    def __init__(self, value):
        self.value = value

    def __repr__(self):
        return f"ResultOk({self.value!r})"


class ResultErr:
    def __init__(self, error):
        self.error = error

    def __repr__(self):
        return f"ResultErr({self.error!r})"


class ResultServiceRt(IBaboonServiceRt):
    def pure(self, value):
        return ResultOk(value)

    def fail(self, error):
        return ResultErr(error)

    def left_map(self, value, f):
        if isinstance(value, ResultErr):
            return ResultErr(f(value.error))
        return value

    def flat_map(self, value, f):
        if isinstance(value, ResultErr):
            return value
        return f(value.value)

File: test/sc-stub-hkt-overlay/src/main/scala/CustomContainers.scala
package custom

sealed trait MyBi[+L, +R]
object MyBi {
  case class Good[+L, +R](value: R) extends MyBi[L, R]
  case class Bad[+L, +R](error: L) extends MyBi[L, R]
}

class MyBiServiceRt extends testpkg.pkg0.IBaboonServiceRt[MyBi] {
  def pure[L, R](value: R): MyBi[L, R] = MyBi.Good(value)
  def fail[L, R](error: L): MyBi[L, R] = MyBi.Bad(error)
  def leftMap[A, B, C](value: MyBi[A, B], f: A => C): MyBi[C, B] = value match {
    case MyBi.Bad(e) => MyBi.Bad(f(e))
    case MyBi.Good(v) => MyBi.Good(v)
  }
  def flatMap[A, B, C](value: MyBi[A, B], f: B => MyBi[A, C]): MyBi[A, C] = value match {
    case MyBi.Bad(e) => MyBi.Bad(e)
    case MyBi.Good(v) => f(v)
  }
}

object MyBiServiceRt {
  val instance: MyBiServiceRt = new MyBiServiceRt()
}
